### ISSUE TITLE
Invalid keyscope name reported

### ISSUE THREAD

**Issue by raducoravu** [dita-ot/dita-ot#2228]
Our QA engineer is French so he defined a @keyscope attribute like:

      <topicref .... keyscope="chrysanthèmes"

which seems to be flagged as invalid by the DITA OT 2.x processing:

    Invalid key scope name "chrysanthèmes".

Looking in the specs, I cannot find a place where the allowed keyscope composing characters are defined.

     https://www.oxygenxml.com/dita/1.3/specs/index.html#langRef/attributes/the-key-scope-attribute.html

**Comment by jelovirt** [comment]
The [DITA 1.3 spec][1] says

> Key names consist of characters that are legal in a URI. The case of key names is significant.

And the [RFC 3986][2] says:

> ... ALPHA (%41-%5A and %61-%7A), DIGIT (%30-%39), hyphen (%2D), period (%2E), underscore (%5F), or tilde (%7E)

Since the DITA spec is not exact here, this can mean either

*   unreserved characters in URI, or
*   data that can be encoded into a URI, for example by UTF-8 encoding it first and then [percent-encoding][3] it.

[1]: http://docs.oasis-open.org/dita/dita/v1.3/os/part3-all-inclusive/langRef/attributes/thekeysattribute.html#thekeysattribute
[2]: http://tools.ietf.org/html/rfc3986#section-2.3
[3]: http://tools.ietf.org/html/rfc3986#section-2.1

**Comment by jelovirt** [comment]
@robander @drmacro @keberlein which is the correct interpretation of the spec?

**Comment by raducoravu** [comment]
The part of the specification you quoted refers to the @keys attribute, there seems to be nothing stated about @keyscope value restrictions.
I just noticed that creating a java.net.URI works over a "chrysanthèmes".

**Comment by jelovirt** [comment]
@raducoravu Yes, you can create a URI with those characters, but those non-ASCII characters are UTF-8 escaped

> A character is encoded by replacing it with the sequence of escaped octets that represent that character in the UTF-8 character set. The Euro currency symbol ('\u20AC'), for example, is encoded as "%E2%82%AC". (Deviation from RFC 2396, which does not specify any particular character set.)

**Comment by jelovirt** [comment]
@raducoravu, quite right, I mixed `@keys` with `@keyscope`. The spec doesn't seem to say anything about `@keyscope` values, just that it's a space separated list of key scopes. So, we can assume it's just "something that has no space characters". Per spec, keyscope can then contain e.g. line feed character since it's not explicitly stated anywhere.

**Comment by raducoravu** [comment]
This is probably just a fault of the specs not stating that the @keyscope should have the same encoding rules as @keys.

**Comment by jelovirt** [comment]
Let's let TC members comment on what is the value space for `@keyscope`. If it's the same as for `@keys`, then we need to clarify what does the spec says/means WRT the "characters that are legal in a URI".

**Comment by drmacro** [comment]
Since key scope names are used in key references they must have the same restrictions.

**Comment by robander** [comment]
I'm surprised that we don't explicitly state what characters are allowed. I'd expect the same limitation that we have on key names. I shudder to think about scope names that include a period. @ctnitchie do you remember any explicit discussion of values allowed as part of a key scope name?

**Comment by jelovirt** [comment]
Note that these key scopes are valid:

*   Unicode REPLACEMENT CHARACTERs, not mangled encoding:

    ```xml
    <topicref keyscope="�����"/>
    ```
*   One key scope that has a line feed and a non-breaking space:

    ```xml
    <topicref keyscope="foo&#xA;bar baz"/>
    ```
*   FULLWIDTH FULL STOP, not a normal dot:

    ```xml
    <topicref keyscope="．"/>
    ```
Nothing wrong with those if authors choose, but I just want to be certain that we use the correct value space to validate the `@keys` and `@keyscope` values.

**Comment by drmacro** [comment]
The language about URIs comes from DITA 1.2 so this ambiguity has always been there.

I'm pretty sure our intent with that language was to limit key names to unescaped characters allowed in URIs.

RFC 3986 says:

"A URI is composed from a limited set of characters consisting of
   digits, letters, and a few graphic symbols.  A reserved subset of
   those characters may be used to delimit syntax components within a
   URI while the remaining characters, including both the unreserved set
   and those reserved characters not acting as delimiters, define each
   component's identifying data."

It is definitely that "limited set of characters" that we intended to be the restriction for key names (and by implication, key scope names).

https://tools.ietf.org/html/rfc3986#page-11

**Comment by jelovirt** [comment]
In order to be as clear as possible, could @dita-ot/tc members actually list all Unicode character ranges that are allowed in key names and key scope names. That is, not reference other specs, but define the value space explicitly.

**Comment by drmacro** [comment]
RFC 3986 defines the following terms, the union of which define the characters that can directly occur in a URI string (that is, the string as evaluated before resolving escaped characters above character 128):

~~~~
   pct-encoded   = "%" HEXDIG HEXDIG

   unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
   reserved      = gen-delims / sub-delims
   gen-delims    = ":" / "/" / "?" / "#" / "[" / "]" / "@"
   sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
                 / "*" / "+" / "," / ";" / "="
~~~~

The terms HEXDIG, ALPHA, and DIGIT are defined in RFC 2234 as:

~~~~
HEXDIG         =  DIGIT / "A" / "B" / "C" / "D" / "E" / "F"
ALPHA          =  %x41-5A / %x61-7A   ; A-Z / a-z
DIGIT          =  %x30-39    ; 0-9
~~~~

I think we can take this to be the full "limited set of characters" since the key names do no explicitly define a way to interpret percent-encoded characters (meaning that key names are not URIs).

The DITA spec explicitly disallows:

 "{", "}", "[", "]", "/", "#", "?" and whitespace (so \u0A and \uA0 are not valid key name characters)

So if we remove the disallowed characters from the RFC 3986 definition, the allowed characters in key names is:

~~~~
 %x41-5A, %x61-7A, "%", "-", ".", "_", "~", ":", "@", "!",  "$",  "&",  "'", "(", ")", "*", "+", ",", ";", "="
 ~~~~

**Comment by jelovirt** [comment]
Does the spec actually specify what "whitespace" characters are?

**Comment by jelovirt** [comment]
The current implementation is https://github.com/dita-ot/dita-ot/blob/develop/src/main/java/org/dita/dost/writer/ValidationFilter.java#L306

Percent cannot be allowed to be an allowed character, since it's the escape character for percent encoding.

**Comment by drmacro** [comment]
I don't see how that follows. 

Key names are not URIs therefore percent encoding has no meaning (because the DITA specification does not otherwise say anything about applying any kind of decoding semantics to key names).

**Comment by jelovirt** [comment]
Yes, but the DITA specification says the characters allowed in key names are those allowed in URIs. In URIs `%` is not allowed character, since it's a reserved character.

**Comment by drmacro** [comment]
% is allowed in URI *strings*, which is what we're talking about, not URIs post URI decoding.

**Comment by drmacro** [comment]
The challenge is what did we mean by "allowed in"? I take it to mean "the undecoded string representation of URIs", in which case all the characters in the ABNF terminals are allowed, which includes the characters in the "reserved" non-terminal.

Because key names are not URIs and are not decoded having "%12" in your key name doesn't hurt anything but it doesn't mean anything either.

**Comment by drmacro** [comment]
Otherwise I think the code you linked to is correct per my analysis of what the DITA spec language implies.

**Comment by georgebina** [comment]
I do not understand the reasoning behind this restriction on key names and implicitly key scope names... Imagine someone from Japan, they will not be able to use familiar names for keys or key scopes. I guess there was an idea that the key names may be used for file names, and thus they can be part of an URL, but then, in this case, the URL could have escaped the characters that cannot appear directly, so not a real problem... Is there a reason why DITA-OT cannot be a little more relaxed on accepting keys and key scopes and maybe just issue a warning in such a case?

**Comment by drmacro** [comment]
It's been long enough (DITA 1.2 time) that I don't remember our deliberations on this issue--maybe Robert has a better memory?

In general I agree with George and my dim memory is that we generally felt the same way: we didn't want to unnecessarily limit the characters that could be used for key names.

On the other hand, I know we (the TC members and tool implementors) tend to think of them as more like XML IDs and so naturally limit the characters we would think to use. Since we don't have any actively-participating members from Asian countries or from countries that don't use latin-based writing systems, nobody pressed on this too hard.

As Jarno points out there are probably practical reasons for disallowing some specific characters, like non-breaking space.

Since the language of the spec is both underspecified and in place since 2010 I'm not sure what we can do definitively at this point.

The fact that nobody considered this a problem in the last 5 years means that either not enough people were using keys or it's not a typical requirement (or both). Which is not to say it's not a problem, just observing that it's taken 5 years for somebody to push this hard on it.

**Comment by georgebina** [comment]
If we want DITA 1.2 keys and 1.3 key scopes to be used in Japan and other countries I guess we need to seriously consider relaxing this, at least from a practical point of view, that is in the DITA-OT processing :).

**Comment by drmacro** [comment]
My main concern would be the implications for combining characters in writing systems where there are several ways to compose the same character (e.g., diacritical marks). I don't think that's much of a concern for CJK languages but could be for European and Middle Eastern languages.

I'll be in Japan starting Monday, I could ask how they feel about using non-latin characters for key names.

**Comment by robander** [comment]
I don't remember too well (Michael was the main voice on the keys/keyref feature for IBM). I have a vague memory that we started out with a list of characters, which then became more complicated with exclusions, and then more complicated, and then finally somebody advocated a return to simplicity by saying "Just use what can go in an URI, with a short list of exceptions."

**Comment by jelovirt** [comment]
@dita-ot/tc Could the whole TC take this up in a call, to raise as an issue for interoperability and clarification to the spec.

**Comment by georgebina** [comment]
If you discuss that, maybe one option will be to allow a name token:
https://www.w3.org/TR/xml11/#NT-Nmtoken

**Comment by keberlein** [comment]
We can do that. In order for me to add this to the agenda, I'd want
    someone to either:
    
      Send an e-mail to the dita-comment list that summarizes what
        the TC needs to discuss and why 
      
      Send an e-mail to the DITA TC list that summarizes what the TC
        needs to discuss and why 
    
    TC members can post to the DITA TC list; non-members can post to
      the dita-comment list.
    Best,
      Kris
      
      Kristen James Eberlein
      Chair, OASIS DITA Technical Committee
      Principal consultant, Eberlein Consulting
      www.eberleinconsulting.com
      +1 919 682-2290; kriseberlein (skype)
      
    
    On 2/26/2016 1:06 AM, Jarno Elovirta
      wrote:
    
    
      @dita-ot/tc
        Could the whole TC take this up in a call, to raise as an issue
        for interoperability and clarification to the spec.
      —
        Reply to this email directly or view
          it on GitHub.

**Comment by drmacro** [comment]
I will bring this to TC

**Comment by drmacro** [comment]
The problem with NMTOKEN exclusively is that it disallows characters that are allowed in URI strings (e.g., the current OT validator), so we can't use NMTOKEN by itself--we'd have to explicitly add those characters that are allowed in URI strings. But I think that might be the best solution overall and it's what I've proposed to the TC.
